src/core/net/ip6.cpp

```diff
@@ -1199,9 +1235,17 @@ otError Ip6::HandleDatagram(Message &aMessage, Netif *aNetif, const void *aLinkM
         }
 #endif
 
-        IgnoreError(ProcessReceiveCallback(aMessage, messageInfo, nextHeader, aFromNcpHost, Message::kCopyToUse));
+        error = ProcessReceiveCallback(aMessage, messageInfo, nextHeader, aFromNcpHost, Message::kCopyToUse);
+
+        if ((error == OT_ERROR_NONE || error == OT_ERROR_NO_ROUTE) && forward &&
+            !ShouldForwardToThread(messageInfo, aFromNcpHost))
```

Comment:
We are potentially calling `ShouldForwardToThread` twice, however calling `ShouldForwardToThread` might be costly. 
Can we optimize that `ShouldForwardToThread` is called only once?
Maybe we can calculate `bool forwardToThread = ShouldForwardToThread(messageInfo, aFromNcpHost)`; just after the `forward` value is determined. Then, we can just use `forwardToThread` instead of calling `ShouldForwardToThread`. 

Thoughts?